### PR TITLE
[web] Position spans absolutely within paragraph

### PR DESCRIPTION
**Pull request by mdebbar**
We previously let HTML's layout control the flow of spans and placeholders inside a paragraph. We just set enough styles to make the browser lay them out the way we want.

This proved challenging, and there were always minor discrepancies between paragraphs rendered to the DOM vs canvas2d. E.g. trailing white spaces, placeholder alignment, etc.

One of the issues that we need to fix is `TextAlign.justify` (https://github.com/flutter/flutter/issues/96533), so it's a good time to try and address the above-mentioned issues as well with the same solution.

The way we are fixing this is by absolutely positioning each span inside the paragraph based on the results of our layout results.

Fixes https://github.com/flutter/flutter/issues/99794
Fixes https://github.com/flutter/flutter/issues/64792

### CODE REVIEW

**Comment by skia-gold** [comment]
Gold has detected about 5 new digest(s) on patchset 1.
View them at https://flutter-engine-gold.skia.org/cl/github/31907

**Comment by skia-gold** [comment]
Gold has detected about 23 new digest(s) on patchset 5.
View them at https://flutter-engine-gold.skia.org/cl/github/31907

File: lib/web_ui/lib/src/engine/text/canvas_paragraph.dart [hunk 1]
```diff
@@ -157,85 +157,44 @@ class CanvasParagraph implements ui.Paragraph {
...
-
-        if (buffer.isNotEmpty) {
-          element.appendText(buffer.toString());
-          buffer.clear();
-        }
 
         if (box is SpanBox) {
-          span = box.span;
-          element = html.document.createElement('span') as html.HtmlElement;
+          lastSpanElement = html.document.createElement('span') as html.HtmlElement;
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/text/canvas_paragraph.dart:178]: Why not `html.SpanElement`?

**Review by yjbanov — APPROVED** [review]
lgtm

**Comment by mdebbar** [comment]
I just confirmed that this PR fixes the Chinese text issue (https://github.com/flutter/flutter/issues/99794).

| Before | After |
| ------ | -----|
| image | image |

File: lib/web_ui/lib/src/engine/text/canvas_paragraph.dart [hunk 2]
```diff
@@ -307,6 +266,9 @@ void _applyNecessaryParagraphStyles({
 }) {
   final html.CssStyleDeclaration cssStyle = element.style;
 
+  // TODO(mdebbar): Now that we absolutely position each span inside the
+  //                paragraph, do we still need these style on <p>?
```
[the remark this answers is not in the document]
Comment (mdebbar) [reply lib/web_ui/lib/src/engine/text/canvas_paragraph.dart:270]: As discussed offline, the classes [`RootStyleNode`](https://github.com/flutter/engine/blob/398a2747c314930c3aa80a10fd4f73b1d0a9aa9e/lib/web_ui/lib/src/engine/text/canvas_paragraph.dart#L574) and [`ChildStyleNode`](https://github.com/flutter/engine/blob/398a2747c314930c3aa80a10fd4f73b1d0a9aa9e/lib/web_ui/lib/src/engine/text/canvas_paragraph.dart#L496) handle the inheritance of styles.

File: lib/web_ui/lib/src/engine/text/canvas_paragraph.dart (hunk 1 above)
Comment (mdebbar) [reply lib/web_ui/lib/src/engine/text/canvas_paragraph.dart:178]: No reason. Is one different from the other?

I'll keep it as `html.document.createElement('span')` for now so I can easily change it to `flt-span` next.

**Comment by skia-gold** [comment]
Gold has detected about 18 new digest(s) on patchset 6.
View them at https://flutter-engine-gold.skia.org/cl/github/31907

**Pull request merged**